database/database.go

```diff
@@ -0,0 +1,151 @@
+package database
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/HouzuoGuo/tiedot/db"
+	"github.com/fatih/structs"
+	"github.com/josephroberts/edge-node-manager/config"
+	"github.com/josephroberts/edge-node-manager/device"
+)
+
+/*
+Uses the tiedot database
+https://github.com/HouzuoGuo/tiedot
+*/
+
+var directory string
+var connection *db.DB
+
+func LoadDevices(uuid string, t *device.Type) (map[int]device.Interface, error) {
+	initialise()
+	query := fmt.Sprintf(`[{"eq": "%s", "in": ["ApplicationUUID"]}]`, uuid)
+	var q interface{}
+	if err := json.Unmarshal([]byte(query), &q); err != nil {
+		return nil, err
+	}
+
+	queryResult := make(map[int]struct{})
+	devices := connection.Use("Devices")
+	if err := db.EvalQuery(q, devices, &queryResult); err != nil {
+		return nil, err
+	}
+
+	result := make(map[int]device.Interface)
+	for key := range queryResult {
+		if device, err := loadDevice(key, t); err != nil {
+			return nil, err
+		} else {
+			result[key] = device
+		}
+	}
+
+	return result, nil
+}
+
+func SaveDevice(d device.Interface) (int, error) {
+	devices := connection.Use("Devices")
+	return devices.Insert(interfaceToMap(d))
+}
+
+func RemoveDevice(key int) error {
+	devices := connection.Use("Devices")
+	return devices.Delete(key)
+}
+
+func UpdateDevices(d map[int]device.Interface) error {
+	initialise()
```

Comment:
See comments below, but not a big fan that this is potentially called every time this function is run. It also makes it hard to reason about when what variables `initialise()` set up are available.